### ISSUE TITLE
Ability to save surface normals also

### ISSUE THREAD

**Issue by janisc** [buddhi1980/mandelbulber2#75]
I plan to use Mandelbulber generated images in a game. It would be cool to use 2d bitmap images with normals so lighting could be faked without any actual geometry.

I didn't find any way to save normals in Mandelbulber. Is it supported? If not is there any change of it coming supported in the future?

**Comment by buddhi1980** [comment]
What data format is used for saving normals? Is this some kind of binary file? I have never seen such a solution. If you give me some reference how it should be implemented, then should be not difficult to add this to Mandelbulber.

**Comment by zebastian** [comment]
@buddhi1980 
we could easily integrate this into the image save system, if png / exr / tiff is suitable, we would only need to get the normals into cimage.
this was also present in old mandelbulber with export to png. Was this some sort of (semi) standard, or did you make this up :smile: ?

**Comment by janisc** [comment]
For the end result: I am trying to accomplish something like in this video but with Mandelbulber generated fractal images:
https://www.youtube.com/watch?v=iPkO1H1JYdw

The background in that video is a simple flat image with no 3D geometry but it can lit and the light behaves realistically because there is a second bitmap containing the surface normals. Of course it requires fixed orientation to the player, but as a background for example it would be sufficient in a game.

For the actual file format, png would be good, with option to choose 8 or 16 bit format.
Normalized normal's X, Y and Z coordinates would be stored in R, B, G components of the normals png file. 0 in color component equals -1 in normalized normal component and 255 (in 8-bit) equals 1 in normalized normal components. Normals png file would be same resolution as the image itself.

Coordinates would need to be in view space in my case, so that any normal that faces directly to the viewer would be xyz(0,0,1) or  xyz(0,0,-1) depending on which kind of coordinate system is used.
I am quite novice for when it comes to 3D so please excuse if my explanation might be a bit confusing and/or plain incorrect. I really appriciate Mandelbulber!

**Comment by zebastian** [comment]
makes totally sense to me, here is how blender makes it:
http://wiki.blender.org/index.php/Doc:2.4/Manual/Textures/Influence/Material/Bump_and_Normal
except for z range, it is almost identical to your idea

**Comment by buddhi1980** [comment]
It looks like doable thing. I think we can implement it in 2.08. For this I need to do some changes in cImage class to do allocation of memory on demand. If we add next buffer for normals, we will increase total memory consumption. We are close to finish 2.07 so that's why I don't want to implement in this release.

**Comment by zebastian** [comment]
now done with this commit: https://github.com/buddhi1980/mandelbulber2/commit/e61e88d29b2fd054e311f547fa261a837068a8e6
will be present with mandelbulber 2.08
@janisc if you want to test this feature you can compile the current code base from source, see here for instructions:
https://github.com/buddhi1980/mandelbulber2/blob/master/mandelbulber2/deploy/README
If you have the program executable run it and under 
File > Program Settings > Image 
you have to enable surface normals.
If you now save the image with Image > "Save as (PNG / JPG / EXR / TIFF)" you will get an image for the normals
![image](https://cloud.githubusercontent.com/assets/7888201/14582213/fe273966-03fe-11e6-8417-d2f1fc24c24e.jpg)
![image_normal](https://cloud.githubusercontent.com/assets/7888201/14582214/fe2be7b8-03fe-11e6-9502-43a02d9c0dc0.jpg)

**Comment by buddhi1980** [comment]
I have tried to render image with normal map, but I'm getting blank *_normal.png file. It's just black. Am I doing something wrong or it doesn't work? I have tried with 8-bit and 16-bit PNG format.

**Comment by zebastian** [comment]
i am not sure, can you run in debug mode?
if so, can you check what happens in render_worker.cpp
around these lines:
```
if(image->GetImageOptional()->optionalNormal)
						image->PutPixelNormal(screenPoint.x + xx, screenPoint.y + yy, normalFloat);
```

**Comment by buddhi1980** [comment]
It sometimes works, sometimes not. I have to find the rule.
By the way, even if it works, normals are not calculated properly. I will look into this part of code.

**Comment by buddhi1980** [comment]
The code is corrected. Now normals are calculated properly.
![imagenormals_normal](https://cloud.githubusercontent.com/assets/11696990/14688426/c8bd6fb0-0742-11e6-831f-769586b5d0f7.png)

I have also found why it didn't worked before. It uses new sImageOptional only when program is restarted, because it's used only in cImage constructor, which is called one on startup. I haven't fixed this yet.

**Comment by buddhi1980** [comment]
In latest commit I have corrected z (blue) element, because it was inverted.

**Comment by zebastian** [comment]
"black image for normal, when normal not activated on program start"
is fixed now with this commit: 91f574fb8d8256f5b91023a47c5e7b6534553308.
new topics regarding imageOptional should move to new issue. 
@buddhi1980 
can you check and close this issue, if OK?
